### PR TITLE
Simple sampling implementation for cartesian points.

### PR DESCRIPTION
**Pull request by shaun-edwards**
This PR will have to be updated one #33 is accepted.  Also, this PR conflicts with #25.  This PR is simply to address a pressing need to perform sampling on Cartesian points.

### CODE REVIEW

**Comment by shaun-edwards** [comment]
@jrgnicho, please review.

File: descartes_core/include/descartes_core/cart_trajectory_pt.h [hunk 1]
```diff
@@ -71,6 +155,9 @@ struct TolerancedFrame: public Frame
   TolerancedFrame(const Frame &a):
     Frame(a) {};
 
+  TolerancedFrame(PositionTolerance pos_tol, OrientationTolerance orient_tol) :
```
Comment (jrgnicho) [inline descartes_core/include/descartes_core/cart_trajectory_pt.h]: Is this constructor really needed, I don't see how it'd be useful to have tolerances on an undefined frame.

File: descartes_core/test/cartesian_robot.cpp [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+
+  bool CartesianRobot::getIK(const Eigen::Affine3d &pose, const std::vector<double> &seed_state,
+                     std::vector<double> &joint_pose) const
+  {
+    bool rtn = false;
+    KDL::Frame kdl_frame;
+    tf::transformEigenToKDL(pose, kdl_frame);
+
+    joint_pose.resize(6, 0.0);
```
Comment (jrgnicho) [inline descartes_core/test/cartesian_robot.cpp:48]: This feels a bit confusing but my interpretation of this is that the joint pose array is being used to hold the cartesian data, if this is the case then will the result be a cartesian pose ? Should "joint_pose" be renamed to something else?

File: descartes_core/test/cartesian_robot.cpp [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+      ROS_WARN_STREAM("Invalid joint pose passed to get FK, joint pose" << joint_pose);
+      rtn = false;
+    }
+
+    return rtn;
+  }
+
+
+
+  bool CartesianRobot::isValid(const std::vector<double> &joint_pose) const
```
Comment (jrgnicho) [inline descartes_core/test/cartesian_robot.cpp:101]: This method isn't part of the original TrajectoryPt interface, it suggests checking a valid joint pose but the code in it checks for cartesian limits.  Should it also be renamed to something more descriptive?

File: descartes_core/test/descartes_core_test/cartesian_robot.h [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  CartesianRobot(double pos_range, double orient_range)  ;
+
+  virtual bool getIK(const Eigen::Affine3d &pose, const std::vector<double> &seed_state,
+                     std::vector<double> &joint_pose) const;
+
+  virtual bool getAllIK(const Eigen::Affine3d &pose, std::vector<std::vector<double> > &joint_poses) const;
+
+  virtual bool getFK(const std::vector<double> &joint_pose, Eigen::Affine3d &pose) const;
+
+  virtual bool isValid(const std::vector<double> &joint_pose) const;
```
Comment (jrgnicho) [inline descartes_core/test/descartes_core_test/cartesian_robot.h:46]: Are these new non interface 'isValid' methods going to be used outside this class?

File: descartes_core/test/cartesian_robot.cpp (hunk 2 above)
Comment (shaun-edwards) [inline descartes_core/test/cartesian_robot.cpp:48]: The Cartesian robot class is a simple test class.  The robot "kinematics" are such that, Cartesian poses map directly to joint poses.  As a result, IK solutions are just a simple mapping from Pose(x,y,z,roll, pitch, yaw) -> Joint[].

File: descartes_core/include/descartes_core/cart_trajectory_pt.h (hunk 1 above)
Comment (shaun-edwards) [inline descartes_core/include/descartes_core/cart_trajectory_pt.h]: You are correct...there should be a `const Eigen::Affine3d &a` argument as well.

File: descartes_core/test/cartesian_robot.cpp (hunk 3 above)
Comment (shaun-edwards) [inline descartes_core/test/cartesian_robot.cpp:101]: `CartesianRobot` implements [robot mode](https://github.com/ros-industrial-consortium/descartes/blob/hydro-devel/descartes_core/include/descartes_core/robot_model.h).  Again, the weird mapping joint<->cartesian is due to the "kinematics"

File: descartes_core/test/descartes_core_test/cartesian_robot.h (hunk 4 above)
Comment (shaun-edwards) [inline descartes_core/test/descartes_core_test/cartesian_robot.h:46]: It is in the [robot_model](https://github.com/ros-industrial-consortium/descartes/blob/hydro-devel/descartes_core/include/descartes_core/robot_model.h#L96) interface.

File: descartes_core/test/cartesian_robot.cpp (hunk 2 above)
Comment (jrgnicho) [inline descartes_core/test/cartesian_robot.cpp:48]: +1

File: descartes_core/test/cartesian_robot.cpp (hunk 3 above)
Comment (jrgnicho) [inline descartes_core/test/cartesian_robot.cpp:101]: +1

File: descartes_core/test/descartes_core_test/cartesian_robot.h (hunk 4 above)
Comment (jrgnicho) [inline descartes_core/test/descartes_core_test/cartesian_robot.h:46]: +1

**Comment by shaun-edwards** [comment]
@jrgnicho, I refactored tolerance frames in `cart_trajectory_pt.h` based on your comments.  Added a test for zero tolerance points (had to refactor the sampling code to address these points).

**Comment by jrgnicho** [comment]
The unit tests failed

```
Running tests...
Test project /home/coky/ros/hydro/cessna_ws/build
    Start 1: _ctest_amos_cessna_support_roslaunch-check_test_roslaunch_test.xml
1/4 Test #1: _ctest_amos_cessna_support_roslaunch-check_test_roslaunch_test.xml ...***Failed    3.40 sec
    Start 2: _ctest_descartes_core_gtest_descartes_core_utest
2/4 Test #2: _ctest_descartes_core_gtest_descartes_core_utest .....................***Failed    0.07 sec
    Start 3: _ctest_descartes_moveit_rostest_test_launch_utest.launch
3/4 Test #3: _ctest_descartes_moveit_rostest_test_launch_utest.launch .............***Failed    1.48 sec
    Start 4: _ctest_kuka_kr210_descartes_rostest_test_launch_utest.launch
4/4 Test #4: _ctest_kuka_kr210_descartes_rostest_test_launch_utest.launch .........***Failed    1.39 sec

0% tests passed, 4 tests failed out of 4

Total Test time (real) =   6.35 sec

The following tests FAILED:
      1 - _ctest_amos_cessna_support_roslaunch-check_test_roslaunch_test.xml (Failed)
      2 - _ctest_descartes_core_gtest_descartes_core_utest (Failed)
      3 - _ctest_descartes_moveit_rostest_test_launch_utest.launch (Failed)
      4 - _ctest_kuka_kr210_descartes_rostest_test_launch_utest.launch (Failed)
Errors while running CTest
```

To generate these tests I ran "catkin_make test", not sure that this is the right way of running them.

**Comment by shaun-edwards** [comment]
try `catkin_make run_tests` to build and run.

**Comment by jrgnicho** [comment]
The tests introduced/modified by this PR work. +1

Something I should mention is that the test for the kuka_kr210_descartes/test/launch/utest.launch fails with a segfault but this issue was not introduced by this PR.

**Pull request merged**